Create a new provisioning example

Patater opened the pull request:
Teach how to manufacture a PSA Crypto device with provisioned keys through a provisioning example that could ostensibly be used in a device factory.

Patater committed:
Add a device provisioning example

Add a device provisioning example that works with two different types of
secure element keys:
    - Keys already stored in the secure element (in physical slot 0 in
      our example), placed there by the secure element vendor or another
      secure facility before coming to the assembly house.
    - Keys generated within the secure element during assembly (into
      physical slot 2 in our example)

The provisiong example fulfils the need post-device-assembly to notify
Mbed Crypto about which keys are present where, so that applications can
use those keys. Mbed Crypto will remember this information across
reboots, so after provisioning, another application can be flashed onto
the device which can use these keys.

Patater committed:
provisioning: Add CSR generation

Automatically create and print a certificate signing request. This makes
it convenient to create device certificates using never-left-the-device,
PSA-managed private keys. Include instructions on how to create and
verify a device certificate in the README.

Patater committed:
provisioning: Add troubleshooting to README

It is a common source of failure that the provisioning data is
overwritten when programming a new flash image. In a new troubleshooting
section in the README, add a recommendation to upgrade the DAPLink
version to a version that supports partial image updates.

gilles-peskine-arm commented on provisioning/README.md:
```diff
@@ -0,0 +1,109 @@
+# Mbed Crypto provisioning example application
```
The name and the short description of the example are misleading. There are three ways you can provision a device onto a trust network:

* Generate a key off-device, import it onto the device, and enter it into the device database.
* Generate a key on-device, export it from the device, and enter it into the device database.
* Already have a key on the device, export it and enter it into the device database.

(Above, “key” can refer to either the private key or its public key in different parts of the same sentence.)

This application covers both on-device generation and an already-existing key, but not import. This is a useful example, but the description does a bad job of setting the context.

The name is misleading because “provisioning” could apply to a key or to a device. Provisioning a key is not covered at all: that would mean importing a key, which the example does not do. Instead the example uses a key that has already been provisioned. The example performs a _part_ of the process of provisioning a device, so it's misleading to call it just “provisioning”. This is an example of simple (challengeless) **on-boarding**.

Patater replied:
The example could do importing, but our ATECC608A driver doesn't yet implement this so I had to remove that bit.

Patater replied:
Another name we use for this sort of application has been "Factory Configuration". I rejected that name because there are multiple factories involved and it wasn't clear which factory was being referenced. We could use with more clear and consistent terminology here.

gilles-peskine-arm commented on provisioning/README.md:
```diff
@@ -0,0 +1,109 @@
+board to make a device. Now, you'd like to be able to use this key from Mbed
+Crypto. How does Mbed Crypto know the key exists? We can tell Mbed Crypto by
+running a provisionining application, which is typically separate from the
+primary application and may be run as a final device assembly step.
```
This paragraph does not do a good job of answering the fundamental introductory questions: what is this about? why should I care? Is this even applicable to me?

> Let's say the secure element manufacturer

Which secure element manufacturer? The audience is application developers. Their focus is on the device as a whole, not on one particular hardware feature.

> You've taken this secure element, paired it with a nice Cortex-M microcontroller on a board to make a device

In many cases, someone has already done that for me: I bought the hardware or the hardware design, and it has a secure element next to the main chip or inside the main chip.

> How does Mbed Crypto know the key exists? We can tell Mbed Crypto by running a provisionining application

A key provisioning application would import a key, not register an existing key. The code here registers an existing key and generates and uploads a new one.

> a provisionining application, which is typically separate from the primary application and may be run as a final device assembly step.

The code that creates and registers the bootstrap key is a device assembly step, but that's not what this example does. The code that uses the bootstrap key to register a production key, which is what this example does, is not an assembly step.

gilles-peskine-arm replied:
Here's how I'd write the introduction.

> The [Microchip ATECC608A](https://www.microchip.com/wwwproducts/en/ATECC608A) is a secure element which can store and use private keys that cannot be extracted. Secure element manufacturers usually create an initial private key as part of the manufacturing process. You can create a new private key as part of the manufacturing process of the whole device, or as part of its deployment. This example application demonstrates how to register an existing key and generate a new key through Mbed Crypto. This application also demonstrates how to create a certificate signing request (CSR) for the keys through Mbed TLS. Once you have a CSR, you can send it to your certificate authority (CA) to obtain a certificate which you'll then use to authenticate the device.

Patater replied:
"The audience is application developers." No. Ordinary Mbed OS application developers should not be doing the device manufacturing. The audience is device manufacturers, people assembling boards and making sure they ship with their firmware in a good state: ready for the ordinary Mbed OS application developer to just run the `mutual-tls` example app or similar.

gilles-peskine-arm commented on provisioning/README.md:
```diff
@@ -0,0 +1,109 @@
+primary application and may be run as a final device assembly step.
+
+
+### Adding pre-provisioned keys
```
While “Adding” is correct, I think it would be less confusing to consistently use “Registering”.

gilles-peskine-arm commented on provisioning/README.md:
```diff
@@ -0,0 +1,109 @@
+Mbed Crypto provides a way to register pre-existing secure element keys:
+`mbedtls_psa_register_se_key()`. This function adds the necessary metadata to
+persistent storage so that the secure element keys can be used by an
+application key ID.
```
> so that a secure element key can be used from an application

or

> so that a secure element key can be used through a key id

gilles-peskine-arm commented on provisioning/README.md:
```diff
@@ -0,0 +1,109 @@
+persistent storage so that the secure element keys can be used by an
+application key ID.
+
+Provide which physical secure element slot the key is in with
```
Why are you suddenly giving me orders? This paragraph is missing a lead like “To provision an existing key,”.

gilles-peskine-arm commented on provisioning/README.md:
```diff
@@ -0,0 +1,109 @@
+persistent storage so that the secure element keys can be used by an
+application key ID.
+
+Provide which physical secure element slot the key is in with
```
I don't think “provide” works here. “Specify”? “Indicate”?

Patater replied:
Provide works, but specify is better

gilles-peskine-arm commented on provisioning/README.md:
```diff
@@ -0,0 +1,109 @@
+
+Provide which physical secure element slot the key is in with
+`psa_set_key_slot_number()`. This function operates on a key attributes
+structure. Complete the filling in of all necessary attributes and then call
```
“Complete the filling in” → “Fill in”. “all necessary attributes” begs the question of which attributes are necessary; if you don't want to go into detail, “any other necessary attribute” would be better. Which attributes are necessary depends on the driver and on the capabilities of the secure element: there may or may not be enough metadata on the secure element to determine the key type and its policy. Since a significant part of the audience is people who will adapt the example to a different secure element, we need to explain this.

gilles-peskine-arm commented on provisioning/README.md:
```diff
@@ -0,0 +1,109 @@
+so forth.
+
+
+### This example requires a certificate authority
```
Ok. Wait what, three certificate authorities? Are running on my device — surely not? And all these files are not on my Mbed OS device, are they?

I don't see anything in this section that's relevant _here_. And what's there is explained in a very confusing way. In real life, `RootCA.key`, `ServerSub.key` and `DeviceSub.key` are not even on the same machine.

Patater replied:
I should move this to the `mutual-tls` README. The provisioning example works fine without any CAs. The CAs are needed for `mutual-tls` and are best covered there.

gilles-peskine-arm commented on provisioning/README.md:
```diff
@@ -0,0 +1,109 @@
+and produce a certificate for the device which you can use in the TLS with
+client authentication example also included in this repository (in the
+`mutual-tls` folder).
+
```
How do I create all the server-side files (`DeviceSub.crt`, `DeviceSub.key`, `DeviceChain.pem`?

Patater replied:
I'm glad you asked!

I was debating adding a section on this in the README. I wasn't sure it had a place here, but it seems obvious now we should have this.

Patater replied:
Will add to the `mutual-tls` example README

gilles-peskine-arm commented on provisioning/README.md:
```diff
@@ -0,0 +1,109 @@
+The Device Sub-CA will consume the device-generated certificate signing request
+and produce a certificate for the device which you can use in the TLS with
+client authentication example also included in this repository (in the
+`mutual-tls` folder).
```
Not found

Patater replied:
Not yet present. I can add this reference back in when the folder is present. Didn't want to add the mutual-tls example in the same PR, but I can add this reference from the provisioning example to the mutual-tls example in the mutual-tls example PR.

gilles-peskine-arm commented on provisioning/mbedtls_user_config.h:
```diff
@@ -0,0 +1,34 @@
+/* Make Mbed TLS use PSA. */
+#define MBEDTLS_USE_PSA_CRYPTO
+
+/* Enable pretty-printing of public keys. */
```
I'd call it “ASCII” or “text” rather than “pretty”.

gilles-peskine-arm commented on provisioning/mbedtls_user_config.h:
```diff
@@ -0,0 +1,34 @@
+/* Enable PSA use of secure elements. */
+#define MBEDTLS_PSA_CRYPTO_SE_C
+
+/* Make Mbed TLS use PSA. */
```
What does that mean? PSA framework? PSA attestation? PSA storage?

> Make Mbed TLS use the PSA Crypto API.

AndrzejKurek replied:
Has this been addressed?

gilles-peskine-arm commented on provisioning/mbedtls_user_config.h:
```diff
@@ -0,0 +1,34 @@
+
+/* Enable PSA APIs, which this example depends on. */
+#define MBEDTLS_PSA_CRYPTO_C
+
```
You also need to enable `MBEDTLS_PSA_CRYPTO_STORAGE_C`.

Patater replied:
We can remove this actually. Mbed OS will set this already.

gilles-peskine-arm reviewed (changes requested):
I've reviewed the documentation so far. More precisely I've reviewed everything except `main.c`. I'll review `main.c` later.

AndrzejKurek commented on provisioning/main.c:
```diff
@@ -0,0 +1,321 @@
+#include "mbed_assert.h"
+
+/* The slot number for the device private key stored in the secure element by
+ * the secure element factory. Note: If, for example, your SE has a
```
Similar comment as Gilles' in a different place - at this stage noone would assume that if there's a pre-provisioned key in slot 0, the slot number cannot be 0.

gilles-peskine-arm replied:
That depends on the reader's background. If they're familiar with the PSA crypto API but not with the secure element interface, they might assume that 0 = invalid/unset, because that's usually true in the API. So it would be worth stating something like “Note that unlike most numerical values in the PSA Crypto API, 0 is a valid value for a slot number. Like many other secure elements, the ATECC608A numbers its slots from 0 to N-1 (and specifically for the ATECC608A, N=16).”

Stating that slot numbers are not key ids belongs in a separate sentence. It's completely unrelated to whether 0 is a valid value.

AndrzejKurek commented on provisioning/main.c:
```diff
@@ -0,0 +1,321 @@
+
+    /* Set device private key attributes. */
+    /* Note: It is not necessary to hard-code the physical slot number within
+     * the secure element. The secure element driver can automatically allocate
```
I understood this sentence a bit differently than intended: hard-coding within the secure element sounds like embedding some kind of a number inside the secure element. Maybe it should be 
`hard-code the number of the physical secure element slot`?

AndrzejKurek commented on provisioning/main.c:
```diff
@@ -0,0 +1,321 @@
+    mbedtls_pk_free(&pk);
+}
+
+/* The secure element factory put a device private key pair (not attestation
```
These comments look more like a draft of reminders for you. What is their purpose?

Patater replied:
Removed the note about the TLS application. Moved the comment to a better place and added a new one.

AndrzejKurek commented on provisioning/main.c:
```diff
@@ -0,0 +1,321 @@
+    free(output);
+}
+
+void generate_csr(psa_key_id_t key_id)
```
This function also prints csr's. Maybe `generate_print_csr`?

gilles-peskine-arm commented on provisioning/main.c:
```diff
@@ -0,0 +1,321 @@
+
+    /* Set device private key attributes. */
+    /* Note: It is not necessary to hard-code the physical slot number within
+     * the secure element. The secure element driver can automatically allocate
```
Why not let it allocate here, since hard-coding the slot number is an unusual use case?

Patater replied:
We haven't implemented that part of our driver yet, as far as I'm aware. Our driver is not smart enough to know which slots can be used for what purpose based on how the chip is configured. We work around this by using a slot we know is configured for our requested use case. This is of course not ideal.

gilles-peskine-arm commented on provisioning/main.c:
```diff
@@ -0,0 +1,321 @@
+    psa_set_key_id(&attributes, EXAMPLE_GENERATED_KEY_ID);
+    psa_set_key_lifetime(&attributes, PSA_ATECC608A_LIFETIME);
+    psa_set_key_algorithm(&attributes, PSA_ALG_ECDSA(PSA_ALG_SHA_256));
+    psa_set_key_bits(&attributes, 256);
```
Minor: it's weird to set the size before the type, since the size is only meaningful for a given type. It's also weird to set the slot number before the lifetime. A better order for the attributes would be:

1. id (or after lifetime and slot number)
1. lifetime
1. slot number
1. type
1. size
1. algorithm
1. usage

(Also applies to `register_preprovisioned_keys`)

gilles-peskine-arm commented on provisioning/main.c:
```diff
@@ -0,0 +1,321 @@
+        puts("Out of memory");
+        return;
+    }
+
```
If you read this program expecting a demonstration of the PSA API, it's a bit weird to see mbedtls API functions here. Please add a comment stating that the PSA API currently lacks a way to format the key in the desired format, so we use the mbedtls pk module.

gilles-peskine-arm commented on provisioning/main.c:
```diff
@@ -0,0 +1,321 @@
+
+void print_public_key(psa_key_id_t key_id)
+{
+    enum { OUTPUT_LEN = 256 };
```
It would be clearer if this constant had a more meaningful name: `PEM_LEN` or `PUBLIC_KEY_PEM_LENGTH`

(Also applies to `generate_csr`)

gilles-peskine-arm commented on provisioning/main.c:
```diff
@@ -0,0 +1,321 @@
+    psa_status_t status;
+    printf("Provisioning device...\n");
+
+    printf("\tErasing device... ");
```
This isn't quite erasing the device — it's not like it's changing the code image. Please clarify what gets erased, and also that this may involve a reboot.

gilles-peskine-arm reviewed (changes requested):
Reviewed `main.c`, completing my [previous review](https://github.com/ARMmbed/mbed-os-example-atecc608a/pull/25#pullrequestreview-316834332).

Patater reviewed (commented):
Rebased to address review feedback

Patater commented:
Updated to address more comments and add some further refinements.

dgreen-arm reviewed (changes requested):
If we want to have this example tested in CI, we'll need a `tests/provisioning.log` file for the CI to compare against. Otherwise we can't test this automatically.

Patater committed:
provisioning: Add expected example output

Add the expected example output so that we can automatically test that
the example is working as intended.

gilles-peskine-arm commented on provisioning/README.md:
```diff
@@ -0,0 +1,81 @@
+Crypto. How does Mbed Crypto know the key exists? We can tell Mbed Crypto by
+running a factory device provisionining application, which is typically
+separate from the primary application and may be run as a final device assembly
+step.
```
After reading this paragraph, I understand that this document is going to show me how to use a pre-provisioned key. So when I reach the section “Generate a new key within a secure element”, I understand it as a step on the path to using a pre-provisioned key, which is puzzling because I don't see why you'd need to generate a key.

Please rewrite this paragraph not to be misleading.

gilles-peskine-arm commented on provisioning/README.md:
```diff
@@ -0,0 +1,81 @@
+For registering keys already present within a secure element, Mbed Crypto
+provides a function to inform Mbed Crypto about the existence of the key:
+`mbedtls_psa_register_se_key()`. This function adds the necessary metadata to
+persistent storage so that the secure element keys can be used by an
```
While this sentence is fine on its own, it's a bit weird to transition from “key” singular to “keys” plural. Also the sentence would be easier to read in the active voice.

gilles-peskine-arm commented on provisioning/README.md:
```diff
@@ -0,0 +1,81 @@
+
+With `psa_set_key_slot_number()`, you can specify which physical secure element
+slot the key is in. This function operates on a key attributes structure. Fill
+in any other necessary attributes and then call `mbedtls_psa_register_se_key()`
```
“any other necessary attributes” makes it seem that just setting the slot number can be enough, but it never is. Lifetime and key id are always necessary, usage policy most of the time. Type and size may or may not be necessary depending on whether the hardware already has this information (e.g. in the slot configuration). Either change it to “other necessary attributes” or list the attributes.

gilles-peskine-arm commented on provisioning/README.md:
```diff
@@ -0,0 +1,81 @@
+number.
+
+For generated keys, unlike pre-existing secure element keys, calling
+`psa_set_key_slot_number()` is not required. The driver will select a valid and
```
It's not required in general, but it's required with this particular driver since this particular driver does not implement slot number allocation.

gilles-peskine-arm commented on provisioning/main.c:
```diff
@@ -0,0 +1,345 @@
+ * secure element by this example provisioning application. */
+#define EXAMPLE_GENERATED_KEY_ID 18
+
+/* Mbed TLS needs a CSPRNG to generate a CSR. Provide it a callback which uses
```
```suggestion
/* Mbed TLS needs a random generator to generate a CSR. Provide it a callback which uses
```
The random generator is used for a random challenge and for blinding. A PRNG would not work for either use.

gilles-peskine-arm commented on provisioning/main.c:
```diff
@@ -0,0 +1,345 @@
+
+    /* Set device private key attributes. */
+    /* Note that it is not necessary to specify the physical slot number within
+     * the secure element. The secure element driver can automatically allocate
```
A driver can do it, but the driver we have now doesn't do it. So it isn't necessary in general, but it's necessary for this particular (version of this particular) driver.

gilles-peskine-arm commented on provisioning/main.c:
```diff
@@ -0,0 +1,345 @@
+    /* We also want to generate our own key entirely within the secure element
+     * during the factory device provisioning stage in device assembly. Ask
+     * Mbed Crypto to generate this key for us. */
+    printf("\tGenerating keys within the secure element... ");
```
Not a blocker, but it's a bit weird that the message says “generating key**s**” but the function is `generate_key` singular. And the registration above uses plural both in the message and in the function name, but actually only registers one key (ok, technically a key pair, but technically a key pair is a single key object).

gilles-peskine-arm commented on tests/provisioning.log:
```diff
@@ -0,0 +1,45 @@
+	Key ID 18:
+-----BEGIN PUBLIC KEY-----
+[0-9a-fA-F+/=]+
+[0-9a-fA-F+/=]+
```
Is there a way to say that there can be an arbitrary number of lines? The exact length of the public key depends on a random factor (it's in a representation that strips leading zeros from numbers). We may be lucky enough that in this particular case the variation is within a range where the number of lines doesn't change, but even so this is brittle.

dgreen-arm replied:
I'm not sure if we can specify an arbitrary number of lines. However, there's this from https://github.com/ARMmbed/mbed-os-tools/tree/master/packages/mbed-host-tests:
```
htrun matches lines from the compare log (on left) one by one.
It keeps on looking for a line until it matches.
Once matched it moves on to match the next line.
```
So if we know what the minimum number of lines the key could be, setting the log to match that many should work in all cases if I'm understanding it right.

gilles-peskine-arm replied:
Oh, good. So as long as the number of matched lines is at most the minimum length of the PEM, which looks plausible, the matching is robust.
